src/rgw/rgw_data_sync.cc

```diff
@@ -5211,11 +5310,21 @@ int RGWSyncBucketCR::operate()
           }
         }
 
-        yield call(new RGWSyncBucketShardCR(sc, data_lease_cr, sync_pair,
-                                            sync_pipe, bucket_status.state,
-                                            tn, progress));
-        if (retcode < 0) {
-          return set_cr_error(retcode);
+      if (sync_pair.source_bs.shard_id >= bucket_status.shards_done_with_gen.size()) {
+        tn->log(1, SSTR("bucket shard " << sync_pair.source_bs << " index out of bounds"));
+        return set_cr_done(); // return success so we don't retry
+      }
+      if (bucket_status.shards_done_with_gen[sync_pair.source_bs.shard_id]) {
+        tn->log(10, SSTR("bucket shard " << sync_pair.source_bs << " of gen " <<
+                        gen << " already synced."));
+        return set_cr_done();
+      }
+
+      yield call(new RGWSyncBucketShardCR(sc, data_lease_cr, sync_pair,
+                                          sync_pipe, bucket_status.state,
+                                          tn, progress));
+      if (retcode < 0) {
+        return set_cr_error(retcode);
```

Comment:
it looks like the indentation is off for this whole block

Reply:
yes it should have been under Incremental sync block. thanks